File: backend/app/schemas/simulator.py
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ScenarioType = Literal[
    "supplier_disruption",
    "country_disruption",
    "commodity_shock",
    "operational_deterioration",
]
SupplierDisruptionSeverity = Literal["moderate", "severe", "unavailable"]
OperationalTargetType = Literal["supplier", "country", "commodity"]


class SimulatorScenarioRequest(BaseModel):
    scenarioType: ScenarioType
    supplierId: int | None = Field(default=None, gt=0)
    severity: SupplierDisruptionSeverity | None = None
    targetType: OperationalTargetType | None = None
    targetValue: str | None = None
    delayIncreasePct: float | None = Field(default=None, ge=0, le=100)
    defectIncreasePct: float | None = Field(default=None, ge=0, le=100)
    costVarianceIncreasePct: float | None = Field(default=None, ge=0, le=100)

    model_config = ConfigDict(from_attributes=True)


class SimulatorKpiSummary(BaseModel):
    highRiskSuppliers: int
    mediumRiskSuppliers: int
    lowRiskSuppliers: int
    avgOverallRisk: float
    avgOperationalRisk: float
    avgEsgRisk: float

    model_config = ConfigDict(from_attributes=True)


class SimulatorDeltaSummary(BaseModel):
    highRiskSuppliers: int
    mediumRiskSuppliers: int
    lowRiskSuppliers: int
    avgOverallRisk: float
    avgOperationalRisk: float
    avgEsgRisk: float

    model_config = ConfigDict(from_attributes=True)


class SimulatorRiskBandMovement(BaseModel):
    fromBand: str
    toBand: str
    supplierCount: int

    model_config = ConfigDict(from_attributes=True)


class SimulatorAffectedSupplierItem(BaseModel):
    supplierId: int
    supplierName: str
    country: str | None
    beforeOverallRisk: float
    afterOverallRisk: float
    deltaOverallRisk: float
    beforeRiskLevel: str
    afterRiskLevel: str
    impactReason: str

    model_config = ConfigDict(from_attributes=True)


class SimulatorScenarioMeta(BaseModel):
    scenarioType: ScenarioType
    title: str
    summary: str
    supplierId: int | None = None
    supplierName: str | None = None
    country: str | None = None
    severity: SupplierDisruptionSeverity | None = None
    targetType: OperationalTargetType | None = None
    targetValue: str | None = None
    delayIncreasePct: float | None = None
    defectIncreasePct: float | None = None
    costVarianceIncreasePct: float | None = None

    model_config = ConfigDict(from_attributes=True)


class SimulatorScenarioResponse(BaseModel):
    scenario: SimulatorScenarioMeta
    before: SimulatorKpiSummary
    after: SimulatorKpiSummary
    deltas: SimulatorDeltaSummary
    riskBandMovement: list[SimulatorRiskBandMovement]
    affectedSuppliers: list[SimulatorAffectedSupplierItem]

    model_config = ConfigDict(from_attributes=True)


class SimulatorOptionItem(BaseModel):
    label: str
    value: str

    model_config = ConfigDict(from_attributes=True)


class SimulatorOptionsResponse(BaseModel):
    countries: list[SimulatorOptionItem]
    commodities: list[SimulatorOptionItem]

    model_config = ConfigDict(from_attributes=True)

File: frontend/src/api/simulator.ts
import { apiRequest } from "./client";

export type ScenarioType =
  | "supplier_disruption"
  | "country_disruption"
  | "commodity_shock"
  | "operational_deterioration";
export type SupplierDisruptionSeverity = "moderate" | "severe" | "unavailable";
export type OperationalTargetType = "supplier" | "country" | "commodity";

export interface SupplierDisruptionScenarioRequest {
  scenarioType: "supplier_disruption";
  supplierId: number;
  severity: SupplierDisruptionSeverity;
}

export interface OperationalDeteriorationScenarioRequest {
  scenarioType: "operational_deterioration";
  targetType: OperationalTargetType;
  targetValue: string;
  delayIncreasePct: number;
  defectIncreasePct: number;
  costVarianceIncreasePct: number;
}

export interface CountryDisruptionScenarioRequest {
  scenarioType: "country_disruption";
  targetValue: string;
  severity: SupplierDisruptionSeverity;
}

export interface CommodityShockScenarioRequest {
  scenarioType: "commodity_shock";
  targetValue: string;
  severity: SupplierDisruptionSeverity;
}

export type SimulatorScenarioRequest =
  | SupplierDisruptionScenarioRequest
  | CountryDisruptionScenarioRequest
  | CommodityShockScenarioRequest
  | OperationalDeteriorationScenarioRequest;

export interface SimulatorKpiSummary {
  highRiskSuppliers: number;
  mediumRiskSuppliers: number;
  lowRiskSuppliers: number;
  avgOverallRisk: number;
  avgOperationalRisk: number;
  avgEsgRisk: number;
}

export interface SimulatorDeltaSummary {
  highRiskSuppliers: number;
  mediumRiskSuppliers: number;
  lowRiskSuppliers: number;
  avgOverallRisk: number;
  avgOperationalRisk: number;
  avgEsgRisk: number;
}

export interface SimulatorRiskBandMovement {
  fromBand: string;
  toBand: string;
  supplierCount: number;
}

export interface SimulatorAffectedSupplierItem {
  supplierId: number;
  supplierName: string;
  country: string | null;
  beforeOverallRisk: number;
  afterOverallRisk: number;
  deltaOverallRisk: number;
  beforeRiskLevel: string;
  afterRiskLevel: string;
  impactReason: string;
}

export interface SimulatorScenarioMeta {
  scenarioType: ScenarioType;
  title: string;
  summary: string;
  supplierId?: number | null;
  supplierName?: string | null;
  country?: string | null;
  severity?: SupplierDisruptionSeverity | null;
  targetType?: OperationalTargetType | null;
  targetValue?: string | null;
  delayIncreasePct?: number | null;
  defectIncreasePct?: number | null;
  costVarianceIncreasePct?: number | null;
}

export interface SimulatorScenarioResponse {
  scenario: SimulatorScenarioMeta;
  before: SimulatorKpiSummary;
  after: SimulatorKpiSummary;
  deltas: SimulatorDeltaSummary;
  riskBandMovement: SimulatorRiskBandMovement[];
  affectedSuppliers: SimulatorAffectedSupplierItem[];
}

export interface SimulatorOptionItem {
  label: string;
  value: string;
}

export interface SimulatorOptionsResponse {
  countries: SimulatorOptionItem[];
  commodities: SimulatorOptionItem[];
}

export async function getSimulatorOptions(): Promise<SimulatorOptionsResponse> {
  return apiRequest<SimulatorOptionsResponse>("/simulator/options");
}

export async function runSimulation(
  payload: SimulatorScenarioRequest,
): Promise<SimulatorScenarioResponse> {
  return apiRequest<SimulatorScenarioResponse>("/simulator/run", {
    method: "POST",
    json: payload,
  });
}
